File: developer signatures/DeveloperSignatureGenerator.php
<?php

define("_AVOCADO_API_URL_BASE", "https://avocado.io/api/");
define("_AVOCADO_API_URL_LOGIN", _AVOCADO_API_URL_BASE . "authentication/login");
define("_AVOCADO_API_URL_COUPLE", _AVOCADO_API_URL_BASE . "couple");
define("_AVOCADO_COOKIE_NAME", "user_email");
define("_AVOCADO_USER_AGENT", "Avocado Test Api Client v.1.0");

// Comment these out if you don't want to use this as a command-line script.
$api = new AvocadoAPI();
$api->updateFromCommandLineInput();

class AvocadoAPI {
  var $couple,
      $authManager;

  function AvocadoAPI() {
    $this->authManager = new AvocadoAuthManager();
  }

  function updateFromCommandLineInput() {
    $this->authManager->updateAuthFromCommandLineInput();
    $this->updateCouple();

    # Check that the response from the Avocado API was valid.
    if ($this->couple == null) {
      print "FAILED.  Signature was tested and failed. Try again and check the auth information.\n";
    } else {
      print "SUCCESS.\n\nBelow is your Avocado API signature:\n" .
        $this->authManager->signature . "\n";
    }
  }

  function updateCouple() {
    # Send the POST request.
    $ch = curl_init(_AVOCADO_API_URL_COUPLE);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    $this->signCurlRequest($ch);
    $output = curl_exec($ch);
    $response_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    # Use the HTTP response code to test if a valid API request was made.
    $this->couple = $response_code == 200 ? json_decode($output) : null;
  }

  function signCurlRequest($ch) {
    curl_setopt($ch, CURLOPT_COOKIE, _AVOCADO_COOKIE_NAME . "=" . $this->authManager->cookie);
    curl_setopt($ch, CURLOPT_HTTPHEADER, array("X-AvoSig: " . $this->authManager->signature));
    curl_setopt($ch, CURLOPT_USERAGENT, _AVOCADO_USER_AGENT);
  }
}


class AvocadoAuthManager {
  var $cookie,
      $developer_id,
      $developer_key,
      $email,
      $password,
      $signature;

  function AvocadoAuthManager() {}

  function updateAuthFromCommandLineInput() {
    # Ask the user for all of the necessary authentication info
    $this->email = get_input("Email of an Avocado account");
    $this->password = get_input_silently("Password");
    $this->developer_id = get_input("Developer ID");
    $this->developer_key = get_input("Developer key");
    $this->updateSignature();
  }

  function updateSignature() {
    # Get a new cookie by logging into Avocado.
    $this->updateLoginCookie();

    # Hash the user token.
    $hashed_user_token = hash("sha256", $this->cookie . $this->developer_key);

    # Store the new signature.
    $this->signature = $this->developer_id . ":" . $hashed_user_token;
  }

  function updateLoginCookie() {
    $fields = array(
      'email'=>urlencode($this->email),
      'password'=>urlencode($this->password)
    );

    # Send the POST request.
    $ch = curl_init(_AVOCADO_API_URL_LOGIN);
    curl_setopt($ch, CURLOPT_HEADER, true);
    curl_setopt($ch, CURLOPT_POST, count($fields));
    curl_setopt($ch, CURLOPT_POSTFIELDS,  get_querystring_from_array($fields));
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_USERAGENT, _AVOCADO_USER_AGENT);
    $header = substr(curl_exec($ch), 0, curl_getinfo($ch, CURLINFO_HEADER_SIZE));
    curl_close($ch);

    # Store the cookie for use in later API requests.
    $this->cookie = get_cookie_from_header($header, _AVOCADO_COOKIE_NAME);
  }
}


#-----------------------------------------------------
# Mama's little helpers: functions we needed for this.
#-----------------------------------------------------

function get_querystring_from_array($fields) {
  foreach($fields as $key=>$value) { $fields_string .= $key.'='.$value.'&'; }
  rtrim($fields_string, '&');
  return $fields_string;
}

function get_cookie_from_header($header, $cookie_name) {
  preg_match('/^Set-Cookie: ' . $cookie_name . '=(.*?);/m', $header, $cookie_array);
  return $cookie_array[1];
}

function get_input_silently($msg){
  # NOTE: stty only works on *nix systems.
  system('stty -echo');
  $input = get_input("Password");
  system('stty echo');
  print "\n";
  return $input;
}

function get_input($msg){
  fwrite(STDOUT, "$msg: ");
  return trim(fgets(STDIN));
}

?>

File: developer signatures/DeveloperSignatureGenerator.py

import cookielib
import getpass
import hashlib
import logging
import json
import urllib
import urllib2

AVOCADO_API_URL_BASE = "https://avocado.io/api/";
AVOCADO_API_URL_LOGIN = AVOCADO_API_URL_BASE + "authentication/login";
AVOCADO_API_URL_COUPLE = AVOCADO_API_URL_BASE + "couple";
AVOCADO_COOKIE_NAME = "user_email";
AVOCADO_USER_AGENT = "Avocado Test Api Client v.1.0";
ERROR_MSG = "FAILED.  Signature was tested and failed. Try again and check the auth information."

class AvocadoAPI(object):
  def __init__(self, auth_client):
    '''
    @type authClient: L{AuthClient}
    '''
    self.auth_client = auth_client
    self.couple = None

  def update_from_command_line(self):
    # Ask the user for all of the necessary authentication info
    self.auth_client.email = raw_input("Email of an Avocado account: ")
    self.auth_client.password = getpass.getpass()
    self.auth_client.dev_id = int(raw_input("Developer ID: "))
    self.auth_client.dev_key = raw_input("Developer key: ")

    self.auth_client.update_signature()
    if self.auth_client.dev_signature is None:
      print ERROR_MSG
      return

    self.update_couple();
    if self.couple is None:
      print ERROR_MSG
    else:
      print "\nBelow is your Avocado API signature:"
      print self.auth_client.dev_signature

  def update_couple(self):
    try:
      cookies = cookielib.CookieJar()
      request = urllib2.Request(
        url = AVOCADO_API_URL_COUPLE,
        headers = {
          "Content-type": "application/x-www-form-urlencoded",
          "User-Agent": AVOCADO_USER_AGENT,
          "X-AvoSig": self.auth_client.dev_signature,
          }
      )
      request.add_header('Cookie',
        '%s=%s' % (AVOCADO_COOKIE_NAME, self.auth_client.cookie_value))
      self.couple = urllib2.urlopen(request)

    except urllib2.URLError, e:
      logging.error(e.read())


class AuthClient(object):
  def __init__(self, email = None, password = None, dev_id = 0, dev_key = None):
    '''
    @type email: C{string}
    @type password: C{string}
    @type dev_id: C{int}
    @type dev_key: C{string}
    '''
    self.email = email
    self.password = password
    self.dev_id = dev_id
    self.dev_key = dev_key
    self.dev_signature = None
    self.cookie_value = None

  def get_cookie_from_login(self):
    params = urllib.urlencode({
      "email": self.email,
      "password": self.password,
    })
    try:
      request = urllib2.Request(
        url = AVOCADO_API_URL_LOGIN,
        data = params,
        headers = {
            "Content-type": "application/x-www-form-urlencoded",
            "User-Agent": AVOCADO_USER_AGENT
            }
        )
      response = urllib2.urlopen(request)

      cookies = cookielib.CookieJar()
      cookies.extract_cookies(response, request)
      for cookie in cookies:
          if cookie.name == AVOCADO_COOKIE_NAME:
               self.cookie_value = cookie.value
               break
    except urllib2.URLError, e:
      logging.error(e.read())

  def hash_signature(self):
    hasher = hashlib.sha256()
    hasher.update(self.cookie_value + self.dev_key)
    self.dev_signature = '%d:%s' % (self.dev_id, hasher.hexdigest())

  def update_signature(self):
    self.get_cookie_from_login()
    if self.cookie_value is not None:
      self.hash_signature()


def main():
  logging.basicConfig(level=logging.DEBUG)
  api = AvocadoAPI(AuthClient())
  api.update_from_command_line()


if __name__ == '__main__':
  main()
